lib/CL/devices/tbb/tbb_scheduler.cc

```diff
@@ -0,0 +1,435 @@
+  scheduler.local_mem_size = (1 + (scheduler.local_mem_size - 1) / MAX_EXTENDED_ALIGNMENT) * MAX_EXTENDED_ALIGNMENT;
+  scheduler.printf_buf_global_ptr = reinterpret_cast<uchar*> (pocl_aligned_malloc (MAX_EXTENDED_ALIGNMENT, scheduler.printf_buf_size * scheduler.num_tbb_threads));
+  scheduler.local_mem_global_ptr = reinterpret_cast<char*> (pocl_aligned_malloc (MAX_EXTENDED_ALIGNMENT, scheduler.local_mem_size * scheduler.num_tbb_threads));
+
+  /* create one meta thread to serve as an async interface thread. */
+  pthread_create (&scheduler.meta_thread, NULL, pocl_tbb_driver_thread, NULL);
+
+  scheduler.grain_size = 0;
+  if (pocl_is_option_set("POCL_TBB_GRAIN_SIZE"))
+    {
+      scheduler.grain_size = pocl_get_int_option("POCL_TBB_GRAIN_SIZE", 1);
+      POCL_MSG_PRINT_GENERAL ("TBB: using a grain size of %u\n", scheduler.grain_size);
+    }
+
+  scheduler.selected_partitioner = NONE;
+  const char *ptr = pocl_get_string_option ("POCL_TBB_PARTITIONER", "");
+  if (strlen (ptr) > 0)
+    {
+      if (strncmp (ptr, "affinity", 8) == 0)
+        {
+          scheduler.selected_partitioner = AFFINITY;
+          POCL_MSG_PRINT_GENERAL ("TBB: using affinity partitioner\n");
+        }
+      else if (strncmp (ptr, "auto", 4) == 0)
+        {
+          scheduler.selected_partitioner = AUTO;
+          POCL_MSG_PRINT_GENERAL ("TBB: using auto partitioner\n");
+        }
+      else if (strncmp (ptr, "simple", 6) == 0)
+        {
+          scheduler.selected_partitioner = SIMPLE;
+          POCL_MSG_PRINT_GENERAL ("TBB: using simple partitioner\n");
+        }
+      else if (strncmp (ptr, "static", 6) == 0)
+        {
+          scheduler.selected_partitioner = STATIC;
+          POCL_MSG_PRINT_GENERAL ("TBB: using static partitioner\n");
+        }
+      else
+        {
+          POCL_MSG_WARN ("TBB: Malformed string in POCL_TBB_PARTITIONER env var: %s\n", ptr);
+        }
+    }
+}
+
+void
+tbb_scheduler_uninit ()
+{
+  unsigned i;
+
+  POCL_FAST_LOCK (scheduler.wq_lock_fast);
+  scheduler.meta_thread_shutdown_requested = 1;
+  pthread_cond_broadcast (&scheduler.wake_meta_thread);
+  POCL_FAST_UNLOCK (scheduler.wq_lock_fast);
+
+  pthread_join (scheduler.meta_thread, NULL);
+
+  POCL_FAST_DESTROY (scheduler.wq_lock_fast);
+  pthread_cond_destroy (&scheduler.wake_meta_thread);
+
+  scheduler.meta_thread_shutdown_requested = 0;
+
+  pocl_aligned_free (scheduler.printf_buf_global_ptr);
+  pocl_aligned_free (scheduler.local_mem_global_ptr);
+}
+
+/* push_command and push_kernel MUST use broadcast and wake up all threads,
+   because commands can be for subdevices (= not all threads) */
+void
+tbb_scheduler_push_command (_cl_command_node *cmd)
+{
+  POCL_FAST_LOCK (scheduler.wq_lock_fast);
+  DL_APPEND (scheduler.work_queue, cmd);
+  pthread_cond_broadcast (&scheduler.wake_meta_thread);
+  POCL_FAST_UNLOCK (scheduler.wq_lock_fast);
+}
+
+/* Internal functions */
+
+class WorkGroupScheduler {
```

Comment:
See c591b07 (same as above). The mess is up to me, because this file is a copy of a C file with a sole inserted C++ class, compiled with a C++ compiler (TBB is a C++ library).